include/geos/math/DD.h

```diff
@@ -0,0 +1,126 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2020 Crunchy Data
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#ifndef GEOS_MATH_DD_H
+#define GEOS_MATH_DD_H
+
+#include <cmath>
+
+namespace geos {
+namespace math { // geos.math
+
+/**
+ * \class DD
+ *
+ * \brief
+ * Wrapper for DoubleDouble higher precision mathematics
+ * operations.
+ */
+class GEOS_DLL DD {
+    private:
+        double SPLIT = 134217729.0; // 2^27+1, for IEEE double
+        double hi;
+        double lo;
+
+        int magnitude(double x) const;
+        int signum() const;
+        DD rint() const;
+
+
+    public:
+        DD(double p_hi, double p_lo) : hi(p_hi), lo(p_lo) {};
+        DD(double x) : hi(x), lo(0.0) {};
+        DD(const DD &dd) : hi(dd.hi), lo(dd.lo) {};
+        DD() : hi(0.0), lo(0.0) {};
+
+        bool operator==(const DD &rhs) const
+        {
+            return hi == rhs.hi && lo == rhs.lo;
+        }
+
+        bool operator!=(const DD &rhs) const
+        {
+            return hi != rhs.hi || lo != rhs.lo;
+        }
+
+        bool operator<(const DD &rhs) const
+        {
+            return (hi < rhs.hi) || (hi == rhs.hi && lo < rhs.lo);
+        }
+
+        bool operator<=(const DD &rhs) const
+        {
+            return (hi < rhs.hi) || (hi == rhs.hi && lo <= rhs.lo);
+        }
+
+        bool operator>(const DD &rhs) const
+        {
+            return (hi > rhs.hi) || (hi == rhs.hi && lo > rhs.lo);
+        }
+
+        bool operator>=(const DD &rhs) const
+        {
+            return (hi > rhs.hi) || (hi == rhs.hi && lo >= rhs.lo);
+        }
+
+        friend DD operator+ (const DD &lhs, const DD &rhs);
+        friend DD operator+ (const DD &lhs, double rhs);
+        friend DD operator- (const DD &lhs, const DD &rhs);
+        friend DD operator- (const DD &lhs, double rhs);
+        friend DD operator* (const DD &lhs, const DD &rhs);
+        friend DD operator* (const DD &lhs, double rhs);
+        friend DD operator/ (const DD &lhs, const DD &rhs);
+        friend DD operator/ (const DD &lhs, double rhs);
+
+        static DD determinant(const DD &x1, const DD &y1, const DD &x2, const DD &y2);
+        static DD determinant(double x1, double y1, double x2, double y2);
+        static DD abs(const DD &d);
+        static DD pow(const DD &d, int exp);
+        static DD trunc(const DD &d);
+
+        bool isNaN() const;
```

Comment:
Many of these methods seem like good candidates for inlining (either by defining them where they're declared, or by using a `.inl` file like much of the older work)

Reply:
Mostly they don't get called except in the tests. Most of the work happens in the arithmetic operators and selfAdd/selfMultiply.

Reply:
Seems like just about all the methods should be inlineable, no?  Would be interesting to see the performance diff.